File: qqbot-domain/src/main/java/com/mieai/qqbot/domain/bot/BotRevision.java
package com.mieai.qqbot.domain.bot;

/** Monotonically increasing optimistic-lock revision for bot configuration. */
public record BotRevision(long value) implements Comparable<BotRevision> {
    private static final BotRevision INITIAL = new BotRevision(1L);

    public BotRevision {
        if (value < 1L) {
            throw new IllegalArgumentException("value must be positive");
        }
    }

    public static BotRevision initial() {
        return INITIAL;
    }

    public static BotRevision of(long value) {
        return value == 1L ? INITIAL : new BotRevision(value);
    }

    public BotRevision next() {
        return new BotRevision(Math.incrementExact(value));
    }

    @Override
    public int compareTo(BotRevision other) {
        return Long.compare(value, other.value);
    }
}

File: qqbot-domain/src/main/kotlin/com/mieai/qqbot/domain/bot/BotRevision.kt
package com.mieai.qqbot.domain.bot


/** Monotonically increasing optimistic-lock revision for bot configuration. */
data class BotRevision(val value: Long) : Comparable<BotRevision> {
    init {
        require(value >= 1L) { "value must be positive" }
    }

    fun next(): BotRevision = BotRevision(Math.incrementExact(value))

    override fun compareTo(other: BotRevision): Int = value.compareTo(other.value)

    companion object {
        private val INITIAL = BotRevision(1L)

        fun initial(): BotRevision = INITIAL

        fun of(value: Long): BotRevision = if (value == 1L) INITIAL else BotRevision(value)
    }
}
